R_scripts/slope_calc_master_script.R

```diff
@@ -0,0 +1,299 @@
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
+# Function to select introns for slope calculation based on various criteria          #
+# Args:                                                                               #
+#   intron_info_file: Data frame containing information on introns                    #
+# Returns:                                                                            #
+#   Data frame filtered based on specific criteria, for downstream slope calculation. #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+select_introns_for_slopecalc <- function(intron_info_file) 
+  {
+  intron_info_file <- intron_info_file %>%
+    filter(overlap_outside_reference != TRUE) %>%      # Exclude introns overlapping outside reference
+    filter(supported_by_samples == 1) %>%              # Include only introns supported by 100% of samples
+    filter(strand_measured != 0) %>%                   # Include only introns where the strand annotation is conclusive +/-
+    filter((start_match_reference == TRUE & end_match_reference == TRUE) |  # Fully match reference intro
+             (start_match_reference == TRUE & end_match_reference == FALSE) | # Match start OR end position of reference intro
+             (start_match_reference == FALSE & end_match_reference == TRUE))
+  return(intron_info_file)  # Return the filtered intron info file
+}
+
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# Function to fit a linear model to the coverage data of the RNAseq to an intron.                                              #
+# Args:                                                                                                        #
+#   chromosome_intron_coverage: List of coverage data for introns.                                                        #
+#   component: Component of the linear model to return (default is "all").                                     #
+# Returns:                                                                                                     #
+#   Data frame with lm info on y-intercept, x-intercept, slope, p-value, and sample filename for each intron.  #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
+compute_lm <- function(coverage_per_intron, 
+                       component = "all") 
+  {
+  if (is.null(coverage_per_intron)) { return(NULL) }  # Return NULL if the input is NULL
+  
+  result <- t(sapply(1:length(coverage_per_intron), function(sample_index) 
+    {
+    fit <- lm(as.numeric(coverage_per_intron[[sample_index]]) ~ seq(1:length(coverage_per_intron[[sample_index]])))  # Fit linear model seq() to give positions to coverage values
```

Comment:
For records on - strand, we need to flip the counting of base pairs (to something like length(coverage_per_intron[[sample_index]]):1)